Library/ViewModels/PledgePaymentMethodsViewModelTests.swift

```diff
@@ -1786,4 +1786,231 @@ final class PledgePaymentMethodsViewModelTests: TestCase {
+
+      // Add a new card
+      self.vm.inputs
+        .paymentSheetDidAdd(
+          clientSecret: "si_fake",
+          paymentMethod: "pm_fake"
+        )
+
+      self.scheduler.run()
+
+      // There are now two cards, one in each section
+      XCTAssertEqual(self.reloadPaymentSheetPaymentMethodsCards.lastValue?.count, 1)
+      XCTAssertEqual(self.reloadPaymentMethodsCards.lastValue?.count, 1)
+
+      // The newly added card is disabled and not selected
+      let newCardData = self.reloadPaymentSheetPaymentMethodsCards.lastValue?.first
+      XCTAssertEqual(newCardData?.isEnabled, false)
+      XCTAssertEqual(newCardData?.isSelected, false)
+
+      // The previously added and available card is still selected
+      XCTAssertEqual(self.reloadPaymentMethodsSelectedCardId.lastValue, UserCreditCards.visa.id)
+      XCTAssertEqual(self.reloadPaymentMethodsIsSelected.lastValue, [true])
+      XCTAssertEqual(self.reloadPaymentMethodsAvailableCardTypes.lastValue, [true])
+    }
+  }
+
+  func testAddNewUnavailablePaymentMethod_userHasNoCards_noCardSelected() {
+    let cards = UserCreditCards.withCards([])
+
+    let graphUser = GraphUser.template |> \.storedCards .~ cards
+    let graphUserResponse = UserEnvelope<GraphUser>(me: graphUser)
+    let addPaymentSheetResponse = CreatePaymentSourceEnvelope.paymentSourceSuccessTemplate
+
+    let mockService = MockService(
+      addPaymentSheetPaymentSourceResult: .success(addPaymentSheetResponse),
+      fetchGraphUserResult: .success(graphUserResponse)
+    )
+    let project = Project.template
+      |> \.availableCardTypes .~ ["VISA"]
+
+    withEnvironment(apiService: mockService, currentUser: User.template) {
+      self.vm.inputs
+        .configure(with: (
+          User.template,
+          project,
+          "checkoutID",
+          Reward.template,
+          .pledge,
+          .discovery
+        ))
+      self.vm.inputs.viewDidLoad()
+
+      self.scheduler.run()
+
+      // There are no cards
+      XCTAssertEqual(self.reloadPaymentMethodsIsSelected.lastValue, [])
+      XCTAssertEqual(self.reloadPaymentMethodsAvailableCardTypes.lastValue, [])
+      XCTAssertNil(self.reloadPaymentMethodsSelectedCardId.lastValue ?? nil)
+
+      // Add a new card
+      self.vm.inputs
+        .paymentSheetDidAdd(
+          clientSecret: "si_fake",
+          paymentMethod: "pm_fake"
+        )
+
+      self.scheduler.run()
+
+      // There is now one card, in the new section
+      XCTAssertEqual(self.reloadPaymentSheetPaymentMethodsCards.lastValue?.count, 1)
+      XCTAssertEqual(self.reloadPaymentMethodsCards.lastValue?.count, 0)
+
+      // The newly added card is disabled and not selected
+      let newCardData = self.reloadPaymentSheetPaymentMethodsCards.lastValue?.first
+      XCTAssertEqual(newCardData?.isEnabled, false)
+      XCTAssertEqual(newCardData?.isSelected, false)
+      XCTAssertNil(self.reloadPaymentMethodsSelectedCardId.lastValue ?? nil)
```

Comment:
Nit: I haven't looked through all the tests again so this may be unnecessary, but should we also test that the pledge button doesn't get enabled in this test?

Reply:
➕

Reply:
Ahh that's a great idea _but_ this test doesn't have the pledge button! It's just the isolated payment sheet. The test for pledges don't have access to the payment sheet itself. So I'm not sure we actually _can_ test the complete behavior of the bug.

Reply:
The `PledgeViewController` watches the signal/delegate call for when a card is selected, so testing that behavior should hopefully be sufficient.